File: Sources/LinuxSupport/ioctl.swift
import LinuxExtensions

@_transparent
public func EVIOCGBIT(_ flag: Int32, _ count: Int32) -> UInt32 {
    return ((2) << (((0 + 8) + 8) + 14)) 
            | ((UInt32(Character("E").asciiValue!)) << (0 + 8)) 
            | ((32 + UInt32(flag)) << 0) 
            | (UInt32(count) << ((0 + 8) + 8))
}

@_transparent
public func ioctl(_ fd: Int32, _ request: Int32, _ ptr: UnsafeMutableRawPointer) -> Int32 {
    return ioctl_ptr(fd, request, ptr)
}

@_transparent
public func ioctl(_ fd: Int32, _ request: Int32, _ value: Int32) -> Int32 {
    return ioctl_value(fd, request, value)
}

File: Sources/LinuxExtensions/ioctl.c
// Game Controllers
#include "Include/ioctl.h"
#include <linux/types.h>

struct input_absinfo {
	int value;
	__s32 minimum;
	__s32 maximum;
	__s32 fuzz;
	__s32 flat;
	__s32 resolution;
};

int ioctl_ptr(int fd, int request, void* ptr) {
    return ioctl(fd, request, ptr);
}
int ioctl_value(int fd, int request, int value) {
    return ioctl(fd, request, value);
}

int EVIOCGBIT(int ev, int len) {
    return _IOC(_IOC_READ, 'E', 0x20 + (ev), len);
}

int EVIOCGABS(int abs) {
     return _IOR('E', 0x40 + (abs), struct input_absinfo);
}
